src/gpu_metal.rs

```diff
@@ -0,0 +1,147 @@
+extern crate metal;
+extern crate objc;
+
+use self::metal::*;
+use self::objc::rc::autoreleasepool;
+use std::mem;
+use byteorder::{ByteOrder, LittleEndian};
+
+static LIBRARY_SRC: &str = include_str!("work.metal");
+static mut LIBRARY: Option<Box<Library>> = None;
+static mut KERNEL: Option<Box<Function>> = None;
+
+pub type Result<T> = std::result::Result<T, String>;
+
+struct GpuTask {
+    difficulty: u64,
+    root: [u8;32]
+}
+
+pub struct Gpu {
+    task: GpuTask,
+    device_idx: usize,
+    threads: usize,
+    threads_per_grid: Option<MTLSize>,
+}
+
+impl Gpu {
+    pub fn new(
+        _platform_idx: usize,
+        device_idx: usize,
+        threads: usize,
+        _local_work_size: Option<usize>,
+    ) -> Result<Gpu> {
+
+        let gpu = Gpu {
+            task: GpuTask {difficulty: 0, root: [0;32]},
+            device_idx: device_idx,
+            threads: threads,
+            threads_per_grid: None,
+        };
+
+        let mut index = 0;
+        let devices = Device::all();
+        for dev in &devices {
+            println!("Metal SDK device {}: {} {}", index, dev.name(), if index == gpu.device_idx {" (selected)"} else {""});
+            index+=1;
+        }
+
+        Ok(gpu)
+    }
+
+    pub fn reset_bufs(&mut self) -> Result<()> {
+        Ok(())
+    }
+
+    pub fn set_task(&mut self, root: &[u8;32], difficulty: u64) -> Result<()> {
+        self.reset_bufs()?;
+        self.task.difficulty = difficulty;
+        self.task.root = *root;
+        Ok(())
+    }
+
+    pub fn try(&mut self, out: &mut [u8;8], attempt: u64) -> Result<bool> {
+        let mut attempt_bytes = [0u8; 8];
+        LittleEndian::write_u64(&mut attempt_bytes, attempt);
+
+        let mut difficulty_bytes = [0u8; 8];
+        LittleEndian::write_u64(&mut difficulty_bytes, self.task.difficulty);
+
+        let mut success = false;
+
+        unsafe {
```

Comment:
What all is unsafe in this block? It's quite a large block and at least a good portion of it doesn't seem to be unsafe. I think explanations of what's unsafe and why it's safe in this context would be helpful.

Reply:
May be able to narrow down the unsafe bits, but it got messy when I initially tried not to initially. This is mostly all FFI code, so I think it's basically an unsafe block of code anyway?

Reply:
Leaving as-is for now, someone with actual Rust skills can surely do a follow-up patch to improve this area 👀 😄